README.md
# dougwithseismic/definitely-not-a-robot 🤖

Be a human pretending to be a robot being a human robot human-being so you can do all those things that humans do, but as a robot. Or something like that ♻. This is Puppeteer but with some extra human imperfections baked in.

![GitHub Badge](https://img.shields.io/github/stars/dougwithseismic/definitely-not-a-robot?style=social&label=Star)

![Definitely Not A Robot](definitely-not-a-robot.gif)

definitely-not-a-robot is a super simple wrapper around the Puppeteer browser that emulates human-like interactions for automation testing or web scraping tasks.

By leveraging puppeteer-extra and its StealthPlugin, it offers enhanced bot detection circumvention so you can keep on scraping and stay one step ahead.

The primary goal of definitely-not-a-robot is to simulate behaviors that are indistinguishable from real human users. This is achieved by introducing randomness in various interactions, making the actions less predictable and more "human-like."

Beep Boop. I don't condone this behaviour.

## Why is randomness important?

In the world of automation, predictability is often the enemy. Web servers, especially those with anti-bot measures, can easily detect repeated patterns typical of bots. By introducing randomness:

- **Mimicking Human Behavior**: Humans are inherently unpredictable. Randomness in delays, mouse movements, and typing speeds makes the browser actions more similar to real users.
  
- **Evading Detection**: Many modern websites employ bot detection tools. Varying the behavior reduces the risk of being flagged as a bot, especially during web scraping or automated browsing.
  
- **Enhanced Test Scenarios**: For testing purposes, unpredictable actions can simulate a broader range of user interactions, potentially uncovering more bugs or issues.

## Features

### Initialization

Start by creating a new instance:

```typescript
import definitelyNotAHuman from 'path-to-definitely-not-a-robot'

const browser = new definitelyNotAHuman()
```

### Launching the browser

Initialize the browser:

```typescript
await browser.launch()
```

For custom configurations, pass the compatible puppeteer's browser launch options:

```typescript
await browser.launch({
  headless: false,
  defaultViewport: { width: 1440, height: 900 }
})
```

Anything you can do with a regular `puppeteer` instance, you can do here.

### Navigating to a URL

Navigate with ease:

```typescript
await browser.navigate('https://www.example.com')
```

### Human-like Mouse Movement to Elements

Move the mouse to an element in a non-linear, humanized path, then click it. To act like a human, you must first move like a human so we split our mouse movement into segments and introduce randomness into each, depending on the distance between the current and target positions.

```typescript
await browser.humanMove('#my-element-id', { hesitationBeforeClick: true })
```

Additional movement options, like `duration` and `steps`, allow for even more varied mouse movement behaviors.

### Simulating Jittery Mouse Movements

Humans rarely keep the mouse perfectly still:

```typescript
await browser.jitterMouse({ jitterCount: 5 })
```

The `jitterMouse` function emulates this behavior by making small, random movements around the current position.

### Typing Text with Human Nuances

Humans don't type at a constant speed:

```typescript
await browser.humanType('Hello, world!', 120) // Typing at approximately 120 words per minute
```

The function simulates human typing patterns by varying speeds between keypresses and occasionally introducing pauses, especially after punctuations. A small delay based on the difference in ASCII values is introduced to add that more flesh-bag human-like flow.

### Randomized Pauses

Humans don't operate in fixed intervals. Emulate this:

```typescript
await browser.wait(1000, 5000) // Waits for a random duration between 1 and 5 seconds
```

### Closing the browser

Cleanly close the browser post interactions:

```typescript
await browser.close()
```

## Changelog

For a list of changes, check out the [Changelog](CHANGELOG.md).

## Author

Doug withSeismic - <doug@withseismic.com> twitter.com/dougiesilkstone


CHANGELOG.md
# Changelog

## [0.0.2] - 2023-08-02

### Added

#### Human-like Mouse Movement to Elements

Move the mouse to an element in a non-linear, humanized path, then click it. To act like a human, you must first move like a human so we split our mouse movement into segments and introduce randomness into each, depending on the distance between the current and target positions.

```typescript
await browser.humanMove('#my-element-id', { hesitationBeforeClick: true })
```


src/index.ts
import HumanBrowser from './HumanBrowser'

const HEADLESS_MODE = false
// const BROWSER_WINDOW_ARGUMENTS = ['--start-maximized'];

const DEFAULT_PAUSE_OPTIONS = {
    maxPause: 2,
    minPause: 0,
    hesitationBeforeClick: true,
}

const JITTER_DEFAULT = {
    jitterMin: 3,
    jitterMax: 30,
    jitterCount: 3,
    debug: true,
}
const URL = `https://ui.shadcn.com`
const SEARCH_BOX_SELECTOR = 'input[type*="search"]'
const TEAM_SELECT_BUTTON = 'button[aria-label="Select a team"]'
const MONSTER_SELECTION = '[data-value="monsters inc."]'
const NAVIGATION_LINK = 'a[href*="/examples/playground"]'
const TEXT_AREA_SELECTOR = 'textarea'

async function typeInSearchBox(browser: HumanBrowser, text: string, wordsPerMinute: number) {
    await browser.humanMove(SEARCH_BOX_SELECTOR, DEFAULT_PAUSE_OPTIONS)
    await browser.jitterMouse(JITTER_DEFAULT)
    await browser.humanType(text, wordsPerMinute)
    await browser.jitterMouse()
}

async function selectTeam(browser: HumanBrowser) {
    await browser.humanMove(TEAM_SELECT_BUTTON, DEFAULT_PAUSE_OPTIONS)
    await browser.jitterMouse()
}

async function typeMonsterName(browser: HumanBrowser, text: string, keyboardStretch: number) {
    await browser.wait(300, 500)
    await browser.jitterMouse({
        jitterMin: 1,
        jitterMax: 30,
        jitterCount: 2,
    })
    await browser.humanType(text, keyboardStretch)
    await browser.wait(300, 500)
    await browser.jitterMouse()
}

async function selectMonster(browser: HumanBrowser) {
    await browser.humanMove(MONSTER_SELECTION, DEFAULT_PAUSE_OPTIONS)
    await browser.jitterMouse()
}

async function navigateToPlayground(browser: HumanBrowser) {
    await browser.humanMove(NAVIGATION_LINK)
    await browser.wait(300, 500)
    await browser.jitterMouse()
}

async function typeTextArea(browser: HumanBrowser, text: string) {
    await browser.humanMove(TEXT_AREA_SELECTOR, {
        target: 'top-left',
        ...DEFAULT_PAUSE_OPTIONS,
    })
    await browser.jitterMouse({
        jitterCount: 1,
    })
    await browser.humanType(text)
}

;(async () => {
    const browser = new HumanBrowser()

    await browser.launch({
        headless: HEADLESS_MODE,
        // args: BROWSER_WINDOW_ARGUMENTS
    })

    await browser.navigate(URL)

    await typeInSearchBox(browser, 'Hello World', 86)
    await selectTeam(browser)
    await typeMonsterName(browser, 'Monste', 30)
    await selectMonster(browser)
    await navigateToPlayground(browser)
    await typeTextArea(browser, `Hello World! I am definitely a human doing human things. How about those student loans and depression, huh? ah ah ah aH AH AHHH.`)

    // ... Continue with other actions ...

    // await browser.close();
})()


src/types.ts
import { Page } from 'puppeteer'

// Options for controlling mouse movement behavior.
export type MovementOptions = {
    scrollBeforeMove?: boolean
    variablePath?: boolean
    hesitationBeforeClick?: boolean
    maxPause?: number
    minPause?: number
    debug?: boolean
    fadeDuration?: number
    startX?: number
    startY?: number
    target?: 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right' | 'center'
    endX?: number
    endY?: number
    jitterMin?: number
    jitterMax?: number
    jitterCount?: number
}

// Function type for debug visualization of mouse positions.
export type DebugShowPositionFunction = (page: Page, x: number, y: number) => Promise<void>

export type DebugOptions = {
    fadeDuration?: number // Duration (in ms) before the debug visuals fade away. If not provided, they won't fade.
}

export interface CustomWindow extends Window {
    debugCanvas: HTMLCanvasElement | undefined
    debugCanvasCtx: CanvasRenderingContext2D
}


export interface MouseState {
    x: number
    y: number
}



export interface Point {
    x: number;
    y: number;
}

export interface SegmentDeltas {
    segmentDxs: number[];
    segmentDys: number[];
}

export interface MovementCalculations {
    speed: number;
    acceleration: number;
}


src/debugUtils.ts
import { Page } from 'puppeteer'
import { CustomWindow, DebugOptions, MovementOptions } from './types'

declare const window: CustomWindow

function createDebugElement(style: Partial<CSSStyleDeclaration>): HTMLElement {
    const elem = document.createElement('div')
    Object.assign(elem.style, style)
    elem.style.zIndex = '999999'
    document.body.appendChild(elem)
    return elem
}

export async function initDebugCanvas(page: Page) {
    return page.evaluate(() => {
        if (window.debugCanvas) {
            return
        }

        const canvas = document.createElement('canvas')
        Object.assign(canvas.style, {
            position: 'fixed',
            top: '0',
            left: '0',
            pointerEvents: 'none',
            zIndex: '999999',
        })

        canvas.width = window.innerWidth
        canvas.height = window.innerHeight
        window.debugCanvas = canvas
        window.debugCanvasCtx = canvas.getContext('2d')!
        document.body.appendChild(canvas)
    })
}

export async function debugShowPosition(page: Page, x: number, y: number, options?: DebugOptions) {
    return page.evaluate(
        (x, y, options) => {
            const dot = createDebugElement({
                position: 'absolute',
                width: '2px',
                height: '2px',
                borderRadius: '50%',
                backgroundColor: 'red',
                transform: 'translate(-50%, -50%)',
                left: `${x}px`,
                top: `${y}px`,
            })

            if (options?.fadeDuration) {
                setTimeout(() => dot.remove(), options.fadeDuration)
            }
        },
        x,
        y,
        options
    )
}

export async function debugDrawSegment(
    page: Page,
    startX: number,
    startY: number,
    endX: number,
    endY: number,
    options: MovementOptions
) {
    await page.evaluate(
        (startX, startY, endX, endY, options) => {
            const ctx = (window as any).debugCanvasCtx
            if (!ctx) return

            ctx.strokeStyle = '#FF0000' // Color of the curve
            ctx.lineWidth = 2

            ctx.beginPath()
            ctx.moveTo(startX, startY)
            ctx.lineTo(endX, endY)
            ctx.stroke()

            if (options.fadeDuration) {
                setTimeout(() => {
                    const padding = 2 // To ensure the line is fully covered
                    const x = Math.min(startX, endX) - padding
                    const y = Math.min(startY, endY) - padding
                    const width = Math.abs(endX - startX) + 2 * padding
                    const height = Math.abs(endY - startY) + 2 * padding

                    ctx.clearRect(x, y, width, height)
                }, options.fadeDuration)
            }
        },
        startX,
        startY,
        endX,
        endY,
        options
    )
}


src/mouseUtils.ts
import { Page, ElementHandle, Point } from 'puppeteer'
import { initDebugCanvas, debugDrawSegment } from './debugUtils'
import { MouseState, MovementCalculations, MovementOptions } from './types'

const defaultOptions: MovementOptions = {
    scrollBeforeMove: true,
    variablePath: true,
    hesitationBeforeClick: true,
    maxPause: 1,
    minPause: 0,
    debug: true,
    fadeDuration: 500,
}

const mouseState: MouseState = {
    x: 0,
    y: 0,
}

const JITTER_DISTANCE = 3
const MIN_JITTER_COUNT = 1
const MAX_JITTER_COUNT = 5

const BUFFER = 20
const MIN_SEGMENTS = 2
const MAX_SEGMENTS = 50
const CLICK_DELAY_MIN = 50
const CLICK_DELAY_MAX = 150

async function randomPause(page: Page, min: number, max: number) {
    const delay = Math.random() * (max - min) + min
    await new Promise((r) => setTimeout(r, delay))
}

function getEndCoordinates(box: any, target: string | undefined): { x: number; y: number } {
    switch (target) {
        case 'top-left':
            return { x: box.x + Math.random() * BUFFER, y: box.y + Math.random() * BUFFER }
        case 'top-right':
            return {
                x: box.x + box.width - Math.random() * BUFFER,
                y: box.y + Math.random() * BUFFER,
            }
        case 'bottom-left':
            return {
                x: box.x + Math.random() * BUFFER,
                y: box.y + box.height - Math.random() * BUFFER,
            }
        case 'bottom-right':
            return {
                x: box.x + box.width - Math.random() * BUFFER,
                y: box.y + box.height - Math.random() * BUFFER,
            }
        default:
            return { x: box.x + Math.random() * box.width, y: box.y + Math.random() * box.height }
    }
}

export async function humanMouseMove(
    page: Page,
    element: ElementHandle<Element>,
    options: MovementOptions = defaultOptions
) {
    options = { ...defaultOptions, ...options }

    const box = await element.boundingBox()
    if (!box) throw new Error('Failed to fetch the bounding box of the element.')

    const viewport = page.viewport()
    if (!viewport) throw new Error('Viewport is not defined.')

    const startX = options.startX ?? Math.random() * viewport.width
    const startY = options.startY ?? Math.random() * viewport.height

    const { x: endX, y: endY } = getEndCoordinates(box, options.target)

    // await drawBezierMovement(page, startX, startY, endX, endY, options)
    await drawBezierMovementWithJitter(page, startX, startY, endX, endY, options)

    if (options.hesitationBeforeClick) {
        await randomPause(page, options.minPause!, options.maxPause!)
    }

    const clickDelay = CLICK_DELAY_MIN + Math.random() * (CLICK_DELAY_MAX - CLICK_DELAY_MIN)
    await page.mouse.click(endX, endY, { delay: clickDelay })

    return { x: endX, y: endY }
}

function calculateDistance(start: Point, end: Point): number {
    return Math.sqrt(Math.pow(end.x - start.x, 2) + Math.pow(end.y - start.y, 2))
}

function calculateSpeed(distance: number): number {
    const BASE_SPEED = 0.03 // you can adjust this as required
    return BASE_SPEED + distance * 0.001
}

function calculateAcceleration(distance: number): number {
    return 0.002 // constant acceleration
}

function computeMovementCalculations(start: Point, end: Point): MovementCalculations {
    const distance = calculateDistance(start, end)
    return {
        speed: calculateSpeed(distance),
        acceleration: calculateAcceleration(distance),
    }
}

function computeBezier(t: number, p0: number, p1: number, p2: number, p3: number): number {
    const u = 1 - t
    const tt = t * t
    const uu = u * u
    return uu * u * p0 + 3 * uu * t * p1 + 3 * u * tt * p2 + tt * t * p3
}

function easeInOutCubic(t: number): number {
    return t < 0.5 ? 4 * t * t * t : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1
}

export async function drawBezierMovement(
    page: Page,
    startX: number,
    startY: number,
    endX: number,
    endY: number,
    options: MovementOptions
): Promise<void> {
    const ctrlPt1X = startX + Math.random() * (endX - startX) * 0.5;
    const ctrlPt1Y = startY + Math.random() * (endY - startY) * 0.5;
    const ctrlPt2X = endX - Math.random() * (endX - startX) * 0.5;
    const ctrlPt2Y = endY - Math.random() * (endY - startY) * 0.5;

    if (options.debug) {
        await initDebugCanvas(page);
    }

    const distance = calculateDistance({x: startX, y: startY}, {x: endX, y: endY});
    const segments = Math.max(MIN_SEGMENTS, Math.min(MAX_SEGMENTS, Math.floor(distance / 10)));
    const { speed, acceleration } = computeMovementCalculations({ x: startX, y: startY }, { x: endX, y: endY });

    let prevX = startX;
    let prevY = startY;
    for (let i = 0; i <= segments; i++) {
        let t = i / segments;
        
        t += speed + 0.5 * acceleration * t * t;
        t = Math.min(t, 1);
        
        const x = computeBezier(t, startX, ctrlPt1X, ctrlPt2X, endX);
        const y = computeBezier(t, startY, ctrlPt1Y, ctrlPt2Y, endY);

        await page.mouse.move(x, y);

        if (options.debug) {
            await debugDrawSegment(page, prevX, prevY, x, y, options);
        }

        prevX = x;
        prevY = y;

        const easedT = easeInOutCubic(t);
        const adjustedPause = options.minPause! + (options.maxPause! - options.minPause!) * easedT;

        await randomPause(page, adjustedPause, adjustedPause + 10);
    }
}

function splitDistance(
    startX: number,
    startY: number,
    endX: number,
    endY: number,
    maxSegmentLength: number
) {
    const distance = Math.sqrt(Math.pow(endX - startX, 2) + Math.pow(endY - startY, 2))
    const segments = Math.ceil(distance / maxSegmentLength)

    const segmentDxs = []
    const segmentDys = []

    let currentX = startX
    let currentY = startY

    for (let i = 1; i <= segments; i++) {
        let nextX = currentX + (endX - startX) / segments
        let nextY = currentY + (endY - startY) / segments

        segmentDxs.push(nextX - currentX)
        segmentDys.push(nextY - currentY)

        currentX = nextX
        currentY = nextY
    }

    return { segmentDxs, segmentDys }
}

async function introduceJitter(page: Page, options: MovementOptions, jitterCount: number) {
    for (let i = 0; i < jitterCount; i++) {
        const angle = Math.random() * 2 * Math.PI
        const distance = JITTER_DISTANCE * Math.random()

        const prevX = mouseState.x
        const prevY = mouseState.y

        const dx = distance * Math.cos(angle)
        const dy = distance * Math.sin(angle)

        mouseState.x = prevX + dx
        mouseState.y = prevY + dy

        await page.mouse.move(mouseState.x, mouseState.y)

        if (options.debug) {
            await debugDrawSegment(page, prevX, prevY, mouseState.x, mouseState.y, options)
        }

        // Random pause between each jitter
        await randomPause(page, 0, 50)

        // Move the mouse back
        mouseState.x = prevX
        mouseState.y = prevY

        await page.mouse.move(prevX, prevY)

        // Random pause between returning from jitter and next jitter
        await randomPause(page, 10, 200)
    }
}

export async function drawBezierMovementWithJitter(
    page: Page,
    startX: number,
    startY: number,
    endX: number,
    endY: number,
    options: MovementOptions | undefined
) {
    options = { ...defaultOptions, ...options }
    mouseState.x = startX
    mouseState.y = startY

    const viewport = page.viewport()
    if (!viewport) throw new Error('Viewport is not defined.')

    const majorMovement =
        Math.abs(endX - startX) > Math.abs(endY - startY) ? 'horizontal' : 'vertical'
    const maxSegmentLength =
        majorMovement === 'horizontal' ? viewport.width * 0.6 : viewport.height * 0.6

    const { segmentDxs, segmentDys } = splitDistance(startX, startY, endX, endY, maxSegmentLength)

    let currentX = startX
    let currentY = startY

    for (let i = 0; i < segmentDxs.length; i++) {
        await drawBezierMovement(
            page,
            currentX,
            currentY,
            currentX + segmentDxs[i],
            currentY + segmentDys[i],
            options
        )

        currentX += segmentDxs[i]
        currentY += segmentDys[i]

        mouseState.x = currentX
        mouseState.y = currentY

        if (i < segmentDxs.length - 1) {
            const jitterCount = Math.floor(
                Math.random() * (MAX_JITTER_COUNT - MIN_JITTER_COUNT + 1) + MIN_JITTER_COUNT
            )
            await introduceJitter(page, options, jitterCount)
        }
    }
}


src/actions/humanType.ts
import { Page, ElementHandle, KeyInput } from 'puppeteer'

const DEFAULT_WPM = 40
const SENTENCE_ENDINGS = ['.', '!', '?']
const PUNCTUATION_DELAY = 200
const ASCII_MULTIPLIER = 2

// Calculates the typing delay for each character based on the given WPM.
function getDelayForCharacter(char: string, previousChar: string, wpm: number): number {
    const averageDelayPerChar = getAverageDelayPerChar(wpm)

    let delay = averageDelayPerChar

    if (previousChar) {
        const diff = Math.abs(char.charCodeAt(0) - previousChar.charCodeAt(0))
        delay += diff * ASCII_MULTIPLIER
    }

    if (SENTENCE_ENDINGS.includes(char)) {
        delay += PUNCTUATION_DELAY
    }

    return delay
}

// Gets the average delay for each character to type based on WPM.
function getAverageDelayPerChar(wpm: number): number {
    const charsPerMinute = wpm * 5
    return 60000 / charsPerMinute // 60000 ms in a minute
}

// Types a character and waits for a delay.
async function typeCharacter(page: Page, char: string, delay: number): Promise<void> {
    await page.keyboard.press(char as KeyInput)
    await new Promise((res) => setTimeout(res, delay))
}

export async function humanType(
    text: string,
    wpm: number = DEFAULT_WPM,
    page: Page
): Promise<void> {
    if (!page) {
        throw new Error('Page is not initialized. Call launch() first.')
    }

    let previousChar = ''

    for (const char of text) {
        const delay = getDelayForCharacter(char, previousChar, wpm)
        await typeCharacter(page, char, delay)
        previousChar = char
    }
}


src/HumanBrowser.ts
import { Browser, Page, BrowserLaunchArgumentOptions, KeyInput } from 'puppeteer'
import puppeteer from 'puppeteer-extra'
import StealthPlugin from 'puppeteer-extra-plugin-stealth'
import { drawBezierMovement, drawBezierMovementWithJitter, humanMouseMove } from './mouseUtils'
import { MovementOptions } from './types'

puppeteer.use(StealthPlugin())

const DEFAULT_VIEWPORT = {
    width: 1280,
    height: 960,
}

export class BrowserNotInitializedError extends Error {
    constructor() {
        super('Browser or page is not initialized. Please launch it first.')
    }
}

export class ElementNotFoundError extends Error {
    constructor(selector: string) {
        super(`Element with selector "${selector}" not found.`)
    }
}

class HumanBrowser {
    private browser?: Browser
    private lastMousePosition: { x: number; y: number } = { x: 0, y: 0 }
    public page?: Page

    constructor(private puppeteerBrowser = puppeteer) {
        this.puppeteerBrowser = puppeteerBrowser
        this.puppeteerBrowser.use(StealthPlugin())
    }

    async launch(options?: BrowserLaunchArgumentOptions) {
        this.browser = await puppeteer.launch({
            headless: false,
            defaultViewport: DEFAULT_VIEWPORT,
            ...options,
        })

        this.page = await this.browser.newPage()
    }

    async navigate(url: string) {
        if (!this.page) {
            throw new Error('Page is not initialized. Call launch() first.')
        }
        await this.page.goto(url)
    }

    async jitterMouse(options: MovementOptions = {}): Promise<void> {
        if (!this.page) {
            throw new Error('Page is not initialized. Call launch() first.')
        }

        options = { ...options, debug: true, fadeDuration: 800 }

        const jitterMin = options.jitterMin ?? 20
        const jitterMax = options.jitterMax ?? 95
        const jitterCount = options.jitterCount ?? 1 // default to one jitter

        for (let i = 0; i < jitterCount; i++) {
            const jitterAmount = Math.random() * (jitterMax - jitterMin) + jitterMin
            const jitterX =
                this.lastMousePosition.x + (Math.random() * jitterAmount - jitterAmount / 2)
            const jitterY =
                this.lastMousePosition.y + (Math.random() * jitterAmount - jitterAmount / 2)

            // Use the drawBezierMovement function to draw the jitter movement
            await drawBezierMovement(
                this.page,
                this.lastMousePosition.x,
                this.lastMousePosition.y,
                jitterX,
                jitterY,
                options
            )

            this.lastMousePosition = { x: jitterX, y: jitterY }

            // Optionally, you can introduce a pause between jitters for more realistic movement
            await new Promise((resolve) => setTimeout(resolve, 100 + Math.random() * 100))
        }
    }

    async humanMove(selector: string, moveOptions?: MovementOptions) {
        if (!this.page) {
            throw new BrowserNotInitializedError()
        }

        const element = await this.page.$(selector)
        if (!element) {
            throw new ElementNotFoundError(selector)
        }

        const boundingBox = await element.boundingBox()
        if (!boundingBox) {
            throw new Error(`Failed to get bounding box for element with selector "${selector}".`)
        }

        let endX = boundingBox.x + boundingBox.width / 2
        let endY = boundingBox.y + boundingBox.height / 2

        const { x: startX, y: startY } = this.lastMousePosition

        const { x, y } = await humanMouseMove(this.page, element, {
            startX,
            startY,
            endX,
            endY,
            ...moveOptions,
        })

        // Remember the new position
        this.lastMousePosition = { x, y }
    }

    async humanType(text: string, wpm: number = 250) {
        if (!this.page) {
            throw new Error('Page is not initialized. Call launch() first.')
        }

        const words = text.split(/\s+/)
        const totalChars = text.length
        const totalWords = words.length
        const totalTimeToType = (totalWords / wpm) * 60 * 1000 // Total time in milliseconds
        const averageDelayPerChar = totalTimeToType / totalChars

        const diffMultiplier = 2

        for (let i = 0; i < text.length; i++) {
            const char = text[i] as KeyInput
            await this.page.keyboard.press(char)

            let delay = averageDelayPerChar

            if (i > 0) {
                // Adjust delay based on the ASCII difference between current and previous character
                const diff = Math.abs(text.charCodeAt(i) - text.charCodeAt(i - 1))
                // Add a variable delay based on ASCII difference, the multiplier can be adjusted
                delay += diff * diffMultiplier
            }

            // Add a longer pause after sentence-ending punctuation
            if (['.', '!', '?'].includes(char)) {
                delay += 200 // 200ms pause after punctuation
            }

            await new Promise((res) => setTimeout(res, delay))
        }
    }

    async wait(minMs: number = 0, maxMs: number = 0): Promise<void> {
        const delay = Math.random() * (maxMs - minMs) + minMs
        await new Promise((r) => setTimeout(r, delay))
    }

    async close(): Promise<void> {
        if (this.browser) {
            await this.browser.close()
        }
    }
}

export default HumanBrowser
